oio/xcute/orchestrator.py

```diff
@@ -375,77 +375,144 @@ def process_reply(self, beanstalkd_job_id, encoded_reply):
+            self.beanstalkd_workers = beanstalkd_workers
 
-            for beanstalkd_addr, beanstalkd \
-                    in all_beanstalkd_with_tubes.iteritems():
-                if beanstalkd_addr not in self.all_beanstalkd:
-                    self.logger.info('Found beanstalkd %s' % beanstalkd_addr)
+            for i in range(self.refresh_time_beanstalkd_workers):
+                if not self.running:
+                    break
+                sleep(1)
 
-                self.all_beanstalkd[beanstalkd_addr] = beanstalkd + ({},)
+        self.logger.info('Exited beanstalkd workers thread')
 
-            sleep(5)
+    def _find_beanstalkd_workers(self):
+        """
+        Find beanstalkd workers by looking at the score,
+        existing tubes and tube statistics.
+        """
+        all_beanstalkd = self.conscience_client.all_services(
+            'beanstalkd')
 
-        self.logger.info('Exited beanstalkd thread')
+        beanstalkd_workers = dict()
+        for beanstalkd_info in all_beanstalkd:
+            try:
+                beanstalkd = self._check_beanstalkd_worker(beanstalkd_info)
+                if not beanstalkd:
+                    continue
 
-    @staticmethod
-    def get_beanstalkd_tubes(beanstalkd_addr):
-        return Beanstalk.from_url('beanstalkd://' + beanstalkd_addr).tubes()
+                beanstalkd.is_broken = False
+                beanstalkd_workers[beanstalkd.addr] = beanstalkd
+            except Exception as exc:
+                self.logger.error('Fail to check beanstalkd: %s', exc)
+        return beanstalkd_workers
 
-    def get_loadbalanced_workers(self):
+    def _check_beanstalkd_worker(self, beanstalkd_info):
+        """
+        Check beanstalkd worker by looking at the score,
+        existing tubes and tube statistics.
+        """
+        beanstalkd_addr = 'beanstalk://' + beanstalkd_info['addr']
+        beanstalkd_score = beanstalkd_info['score']
+        if beanstalkd_score == 0:
+            self.logger.info(
+                'Ignore beanstalkd %s: score=0', beanstalkd_addr)
+            return None
+
+        new_beanstalkd = False
+        beanstalkd = self.beanstalkd_workers.get(beanstalkd_addr)
+        if not beanstalkd:
+            new_beanstalkd = True
+            beanstalkd = Beanstalk.from_url(beanstalkd_addr)
+            beanstalkd.addr = beanstalkd_addr
+        beanstalkd.score = beanstalkd_score
+
+        beanstalkd_tubes = beanstalkd.tubes()
+        if self.beanstalkd_workers_tube not in beanstalkd_tubes:
+            self.logger.info(
+                'Ignore beanstalkd %s: '
+                'No worker has ever listened to the tube', beanstalkd_addr)
+            return None
+
+        current_stats = beanstalkd.stats_tube(
+            self.beanstalkd_workers_tube)
+        beanstalkd_jobs_ready = current_stats['current-jobs-ready']
+        if beanstalkd_jobs_ready > 0:
+            if current_stats['current-jobs-reserved'] <= 0:
```

Comment:
Can't this happen if the worker has just finished a job and hasn't taken another one yet?

Reply:
In this case, we will (wrongly) ignore this beanstalkd. But this case has a fairly low probability of happening with the configuration of the event agent for xcute (workers = 2, concurrency = 5).

Reply:
Right, and there's a log at least, so I'm ok :+1:
Maybe just add in the log the jobs-ready, jobs-reserved counts though!